(possible) bug with retrying tool downloads

**Describe the bug**

It appears that certain types of failures (tcp rather than http) when downloading tools will not trigger retries and instead lead to a process crash.

**Pants version**
`2.14.0rc1+gitf06bf22b`

**OS**
linux

**Additional info**

This appears to be happening in the bootstrap phase and here is the entirety of the log output:

```
16:27:01.19 [WARN] DEPRECATED: option 'repos' in scope 'python-repos' is scheduled to be removed in version 3.0.0.dev0.

A deprecated alias for `[python-repos].find_links`.
16:27:01.31 [INFO] Starting: Resolving plugins: toolchain.pants.plugin==0.22.0
16:27:15.75 [INFO] Completed: Resolving plugins: toolchain.pants.plugin==0.22.0
16:27:16.33 [INFO] [auth-plugin] Successfully loaded Toolchain token from env var 'TOOLCHAIN_AUTH_TOKEN', expiration: 2022-12-20T20:52:09+00:00.
Exception caught: (pants.engine.internals.scheduler.ExecutionError)
  File "/home/runner/.pants/setup/bootstrap-Linux-x86_64/2.14.0rc1+gitf06bf22b_py38/bin/pants", line 8, in <module>
    sys.exit(main())
  File "/home/runner/.pants/setup/bootstrap-Linux-x86_64/2.14.0rc1+gitf06bf22b_py38/lib/python3.8/site-packages/pants/bin/pants_loader.py", line 112, in main
    PantsLoader.main()
  File "/home/runner/.pants/setup/bootstrap-Linux-x86_64/2.14.0rc1+gitf06bf22b_py38/lib/python3.8/site-packages/pants/bin/pants_loader.py", line 108, in main
    cls.run_default_entrypoint()
  File "/home/runner/.pants/setup/bootstrap-Linux-x86_64/2.14.0rc1+gitf06bf22b_py38/lib/python3.8/site-packages/pants/bin/pants_loader.py", line 90, in run_default_entrypoint
    exit_code = runner.run(start_time)
  File "/home/runner/.pants/setup/bootstrap-Linux-x86_64/2.14.0rc1+gitf06bf22b_py38/lib/python3.8/site-packages/pants/bin/pants_runner.py", line 99, in run
    runner = LocalPantsRunner.create(
  File "/home/runner/.pants/setup/bootstrap-Linux-x86_64/2.14.0rc1+gitf06bf22b_py38/lib/python3.8/site-packages/pants/bin/local_pants_runner.py", line 158, in create
    specs = calculate_specs(
  File "/home/runner/.pants/setup/bootstrap-Linux-x86_64/2.14.0rc1+gitf06bf22b_py38/lib/python3.8/site-packages/pants/init/specs_calculator.py", line 72, in calculate_specs
    (changed_addresses,) = session.product_request(
  File "/home/runner/.pants/setup/bootstrap-Linux-x86_64/2.14.0rc1+gitf06bf22b_py38/lib/python3.8/site-packages/pants/engine/internals/scheduler.py", line 577, in product_request
    return self.execute(request)
  File "/home/runner/.pants/setup/bootstrap-Linux-x86_64/2.14.0rc1+gitf06bf22b_py38/lib/python3.8/site-packages/pants/engine/internals/scheduler.py", line 521, in execute
    self._raise_on_error([t for _, t in throws])
  File "/home/runner/.pants/setup/bootstrap-Linux-x86_64/2.14.0rc1+gitf06bf22b_py38/lib/python3.8/site-packages/pants/engine/internals/scheduler.py", line 498, in _raise_on_error
    raise ExecutionError(

Exception message: 1 Exception encountered:

Engine traceback:
  in select
  in pants.vcs.changed.find_changed_owners
  in pants.backend.project_info.dependees.find_dependees
  in pants.backend.project_info.dependees.map_addresses_to_dependees
  in pants.engine.internals.graph.resolve_dependencies (src/infra/pipeline/tools/vcfdiff.sh:shell_sources)
  in pants.backend.shell.dependency_inference.infer_shell_dependencies
  in pants.backend.shell.dependency_inference.parse_shell_imports
  in pants.core.util_rules.external_tool.download_external_tool
  in downloaded_file
Traceback (no traceback):
  <pants native internals>
Exception: Error reading URL fetch response: request or response body error: error reading a body from connection: connection reset
NoneType: None
FATAL: exception not rethrown
build/pants/ci/common.sh: line 14:   543 Aborted                 (core dumped) ./pants "${pants_args[@]}" 2>&1
       544 Done                    | tee "${PANTS_LOG}"
Error: Tests failed ([47](https://github.com/color/color/actions/runs/3291293838/jobs/5425264218#step:9:48)s)
See https://www.pantsbuild.org/docs/python-test-goal for info on how pants runs python tests.
Run './pants test <path...>' to test file(s) with pants locally.
Error: Process completed with exit code 1.
```

--- a/src/python/pants/engine/fs_test.py
+++ b/src/python/pants/engine/fs_test.py
@@ -1065,18 +1065,21 @@ def test_write_digest_workspace(rule_runner: RuleRunner) -> None:
     assert path2.read_text() == "goodbye"
 
 
+@dataclass(frozen=True)
+class DigestRequest:
+    create_digest: CreateDigest
+
+
+class WorkspaceGoalSubsystem(GoalSubsystem):
+    name = "workspace-goal"
+
+
+class WorkspaceGoal(Goal):
+    subsystem_cls = WorkspaceGoalSubsystem
+    environment_behavior = Goal.EnvironmentBehavior.LOCAL_ONLY
+
+
 def test_workspace_in_goal_rule() -> None:
-    class WorkspaceGoalSubsystem(GoalSubsystem):
-        name = "workspace-goal"
-
-    class WorkspaceGoal(Goal):
-        subsystem_cls = WorkspaceGoalSubsystem
-        environment_behavior = Goal.EnvironmentBehavior.LOCAL_ONLY
-
-    @dataclass(frozen=True)
-    class DigestRequest:
-        create_digest: CreateDigest
-
     @rule
     def digest_request_singleton() -> DigestRequest:
         fc = FileContent(path="a.txt", content=b"hello")

--- a/src/python/pants/engine/fs_test.py
+++ b/src/python/pants/engine/fs_test.py
@@ -1,5 +1,8 @@
 # Copyright 2015 Pants project contributors (see CONTRIBUTORS.md).
 # Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
 import hashlib
 import os
 import pkgutil
@@ -857,6 +860,40 @@ class StubHandler(BaseHTTPRequestHandler):
         self.end_headers()
 
 
+def stub_erroring_handler(error_count_value: int) -> type[BaseHTTPRequestHandler]:
+    """Return a handler that errors once mid-download before succeeding for the next GET.
+
+    This function returns an anonymous class so that each call can create a new instance with its
+    own error counter.
+    """
+    error_num = 1
+
+    class StubErroringHandler(BaseHTTPRequestHandler):
+        error_count = error_count_value
+        response_text = b"Hello, client!"
+
+        def do_HEAD(self):
+            self.send_headers()
+
+        def do_GET(self):
+            self.send_headers()
+            nonlocal error_num
+            if error_num <= self.error_count:
+                msg = f"Returning error {error_num}"
+                error_num += 1
+                raise Exception(msg)
+            self.wfile.write(self.response_text)
+
+        def send_headers(self):
+            code = 200 if self.path == "/file.txt" else 404
+            self.send_response(code)
+            self.send_header("Content-Type", "text/utf-8")
+            self.send_header("Content-Length", f"{len(self.response_text)}")
+            self.end_headers()
+
+    return StubErroringHandler
+
+
 DOWNLOADS_FILE_DIGEST = FileDigest(
     "8fcbc50cda241aee7238c71e87c27804e7abc60675974eaf6567aa16366bc105", 14
 )
@@ -884,6 +921,24 @@ def test_download_missing_file(downloads_rule_runner: RuleRunner) -> None:
                 Snapshot, [DownloadFile(f"http://localhost:{port}/notfound", DOWNLOADS_FILE_DIGEST)]
             )
     assert "404" in str(exc.value)
+
+
+def test_download_body_error_retry(downloads_rule_runner: RuleRunner) -> None:
+    with http_server(stub_erroring_handler(1)) as port:
+        snapshot = downloads_rule_runner.request(
+            Snapshot, [DownloadFile(f"http://localhost:{port}/file.txt", DOWNLOADS_FILE_DIGEST)]
+        )
+    assert snapshot.files == ("file.txt",)
+    assert snapshot.digest == DOWNLOADS_EXPECTED_DIRECTORY_DIGEST
+
+
+def test_download_body_error_retry_eventually_fails(downloads_rule_runner: RuleRunner) -> None:
+    # Returns one more error than the retry will allow.
+    with http_server(stub_erroring_handler(5)) as port:
+        with pytest.raises(Exception):
+            _ = downloads_rule_runner.request(
+                Snapshot, [DownloadFile(f"http://localhost:{port}/file.txt", DOWNLOADS_FILE_DIGEST)]
+            )
 
 
 def test_download_wrong_digest(downloads_rule_runner: RuleRunner) -> None:

--- a/src/rust/engine/src/downloads.rs
+++ b/src/rust/engine/src/downloads.rs
@@ -1,13 +1,14 @@
 // Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
 // Licensed under the Apache License, Version 2.0 (see LICENSE).
 
-use std::io::Write;
+use std::io::{self, Write};
 use std::pin::Pin;
 use std::sync::Arc;
 
 use async_trait::async_trait;
 use bytes::{BufMut, Bytes};
 use futures::stream::StreamExt;
+use hashing::Digest;
 use humansize::{file_size_opts, FileSize};
 use reqwest::Error;
 use tokio_retry::strategy::{jitter, ExponentialBackoff};
@@ -16,6 +17,19 @@ use url::Url;
 
 use crate::context::Core;
 use workunit_store::{in_workunit, Level};
+
+enum StreamingError {
+  Retryable(String),
+  Permanent(String),
+}
+
+impl From<StreamingError> for String {
+  fn from(err: StreamingError) -> Self {
+    match err {
+      StreamingError::Retryable(s) | StreamingError::Permanent(s) => s,
+    }
+  }
+}
 
 #[async_trait]
 trait StreamingDownload: Send {
@@ -27,41 +41,36 @@ struct NetDownload {
 }
 
 impl NetDownload {
-  async fn start(core: &Arc<Core>, url: Url, file_name: String) -> Result<NetDownload, String> {
-    let try_download = || async {
-      core
+  async fn start(
+    core: &Arc<Core>,
+    url: Url,
+    file_name: String,
+  ) -> Result<NetDownload, StreamingError> {
+    let response = core
       .http_client
       .get(url.clone())
       .send()
       .await
-      .map_err(|err| (format!("Error downloading file: {}", err), true))
+      .map_err(|err| StreamingError::Retryable(format!("Error downloading file: {err}")))
       .and_then(|res|
         // Handle common HTTP errors.
         if res.status().is_server_error() {
-          Err((format!(
+          Err(StreamingError::Retryable(format!(
             "Server error ({}) downloading file {} from {}",
             res.status().as_str(),
             file_name,
             url,
-          ), true))
+          )))
         } else if res.status().is_client_error() {
-          Err((format!(
+          Err(StreamingError::Permanent(format!(
             "Client error ({}) downloading file {} from {}",
             res.status().as_str(),
             file_name,
             url,
-          ), false))
+          )))
         } else {
           Ok(res)
-        })
-    };
-
-    // TODO: Allow the retry strategy to be configurable?
-    // For now we retry after 10ms, 100ms, 1s, and 10s.
-    let retry_strategy = ExponentialBackoff::from_millis(10).map(jitter).take(4);
-    let response = RetryIf::spawn(retry_strategy, try_download, |err: &(String, bool)| err.1)
-      .await
-      .map_err(|(err, _)| err)?;
+        })?;
 
     let byte_stream = Pin::new(Box::new(response.bytes_stream()));
     Ok(NetDownload {
@@ -86,12 +95,18 @@ struct FileDownload {
 }
 
 impl FileDownload {
-  async fn start(path: &str, file_name: String) -> Result<FileDownload, String> {
+  async fn start(path: &str, file_name: String) -> Result<FileDownload, StreamingError> {
     let file = tokio::fs::File::open(path).await.map_err(|e| {
-      format!(
+      let msg = format!(
         "Error ({}) opening file at {} for download to {}",
         e, path, file_name
-      )
+      );
+      // Fail quickly for non-existent files.
+      if e.kind() == io::ErrorKind::NotFound {
+        StreamingError::Permanent(msg)
+      } else {
+        StreamingError::Retryable(msg)
+      }
     })?;
     let stream = tokio_util::io::ReaderStream::new(file);
     Ok(FileDownload { stream })
@@ -109,25 +124,72 @@ impl StreamingDownload for FileDownload {
   }
 }
 
-async fn start_download(
+async fn attempt_download(
   core: &Arc<Core>,
-  url: Url,
+  url: &Url,
   file_name: String,
-) -> Result<Box<dyn StreamingDownload>, String> {
-  if url.scheme() == "file" {
-    if let Some(host) = url.host_str() {
-      return Err(format!(
-        "The file Url `{}` has a host component. Instead, use `file:$path`, \
-        which in this case might be either `file:{}{}` or `file:{}`.",
-        url,
-        host,
-        url.path(),
-        url.path(),
-      ));
+  expected_digest: Digest,
+) -> Result<(Digest, Bytes), StreamingError> {
+  let mut response_stream: Box<dyn StreamingDownload> = {
+    if url.scheme() == "file" {
+      if let Some(host) = url.host_str() {
+        return Err(StreamingError::Permanent(format!(
+          "The file Url `{}` has a host component. Instead, use `file:$path`, \
+          which in this case might be either `file:{}{}` or `file:{}`.",
+          url,
+          host,
+          url.path(),
+          url.path(),
+        )));
+      }
+      Box::new(FileDownload::start(url.path(), file_name).await?)
+    } else {
+      Box::new(NetDownload::start(core, url.clone(), file_name).await?)
     }
-    return Ok(Box::new(FileDownload::start(url.path(), file_name).await?));
+  };
+
+  struct SizeLimiter<W: std::io::Write> {
+    writer: W,
+    written: usize,
+    size_limit: usize,
   }
-  Ok(Box::new(NetDownload::start(core, url, file_name).await?))
+
+  impl<W: std::io::Write> Write for SizeLimiter<W> {
+    fn write(&mut self, buf: &[u8]) -> Result<usize, std::io::Error> {
+      let new_size = self.written + buf.len();
+      if new_size > self.size_limit {
+        Err(std::io::Error::new(
+          std::io::ErrorKind::InvalidData,
+          "Downloaded file was larger than expected digest",
+        ))
+      } else {
+        self.written = new_size;
+        self.writer.write_all(buf)?;
+        Ok(buf.len())
+      }
+    }
+
+    fn flush(&mut self) -> Result<(), std::io::Error> {
+      self.writer.flush()
+    }
+  }
+
+  let mut hasher = hashing::WriterHasher::new(SizeLimiter {
+    writer: bytes::BytesMut::with_capacity(expected_digest.size_bytes).writer(),
+    written: 0,
+    size_limit: expected_digest.size_bytes,
+  });
+
+  while let Some(next_chunk) = response_stream.next().await {
+    let chunk = next_chunk.map_err(|err| {
+      StreamingError::Retryable(format!("Error reading URL fetch response: {err}"))
+    })?;
+    hasher.write_all(&chunk).map_err(|err| {
+      StreamingError::Retryable(format!("Error hashing/capturing URL fetch response: {err}"))
+    })?;
+  }
+  let (digest, bytewriter) = hasher.finish();
+  Ok((digest, bytewriter.writer.into_inner().freeze()))
 }
 
 pub async fn download(
@@ -148,49 +210,15 @@ pub async fn download(
         .unwrap()
     )),
     |_workunit| async move {
-      let mut response_stream = start_download(&core2, url, file_name).await?;
-      struct SizeLimiter<W: std::io::Write> {
-        writer: W,
-        written: usize,
-        size_limit: usize,
-      }
-
-      impl<W: std::io::Write> Write for SizeLimiter<W> {
-        fn write(&mut self, buf: &[u8]) -> Result<usize, std::io::Error> {
-          let new_size = self.written + buf.len();
-          if new_size > self.size_limit {
-            Err(std::io::Error::new(
-              std::io::ErrorKind::InvalidData,
-              "Downloaded file was larger than expected digest",
-            ))
-          } else {
-            self.written = new_size;
-            self.writer.write_all(buf)?;
-            Ok(buf.len())
-          }
-        }
-
-        fn flush(&mut self) -> Result<(), std::io::Error> {
-          self.writer.flush()
-        }
-      }
-
-      let mut hasher = hashing::WriterHasher::new(SizeLimiter {
-        writer: bytes::BytesMut::with_capacity(expected_digest.size_bytes).writer(),
-        written: 0,
-        size_limit: expected_digest.size_bytes,
-      });
-
-      while let Some(next_chunk) = response_stream.next().await {
-        let chunk =
-          next_chunk.map_err(|err| format!("Error reading URL fetch response: {}", err))?;
-        hasher
-          .write_all(&chunk)
-          .map_err(|err| format!("Error hashing/capturing URL fetch response: {}", err))?;
-      }
-      let (digest, bytewriter) = hasher.finish();
-      let res: Result<_, String> = Ok((digest, bytewriter.writer.into_inner().freeze()));
-      res
+      // TODO: Allow the retry strategy to be configurable?
+      // For now we retry after 10ms, 100ms, 1s, and 10s.
+      let retry_strategy = ExponentialBackoff::from_millis(10).map(jitter).take(4);
+      RetryIf::spawn(
+        retry_strategy,
+        || attempt_download(&core2, &url, file_name.clone(), expected_digest),
+        |err: &StreamingError| matches!(err, StreamingError::Retryable(_)),
+      )
+      .await
     }
   )
   .await?;

